File: src/main/go/leetcode/lists/offerII/Offer50.go
package offerII

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/16 17:17
 */

//[-2,null,-3]
//-5

func pathSum(root *TreeNode, targetSum int) int {
	if root == nil {
		return 0
	}
	var dfs func(node *TreeNode, sum, targetSum int) int
	dfs = func(node *TreeNode, sum, targetSum int) int {
		var res = 0
		if node == nil {
			return 0
		}

		if sum+node.Val == targetSum {
			res++
		}
		res += dfs(node.Left, sum+node.Val, targetSum)
		res += dfs(node.Right, sum+node.Val, targetSum)
		return res
	}
	return dfs(root, 0, targetSum) + pathSum(root.Left, targetSum) + pathSum(root.Right, targetSum)
}

File: src/main/java/leetcode/lists/offerII/Offer50.java
package leetcode.lists.offerII;

import common.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/7/25 16:42
 */
public class Offer50 {


    int cnt = 0;

    public int pathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return 0;
        }
        rootSum(root, targetSum);
        pathSum(root.left, targetSum);
        pathSum(root.right, targetSum);
        return cnt;
    }

    public void rootSum(TreeNode root, long targetSum) {
        if (root == null) {
            return;
        }

        if (root.val == targetSum) {
            cnt++;
        }
        rootSum(root.left, targetSum - root.val);
        rootSum(root.right, targetSum - root.val);
    }
}
